pkg/aspect/types.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aspect
+
+import (
+	"io"
+
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/genproto/googleapis/rpc/code"
+
+	"istio.io/mixer/pkg/attribute"
+	"istio.io/mixer/pkg/expr"
+
+	istiopb "istio.io/api/istio/config/v1"
+)
+
+type (
+	// Adapter represents a factory to create an adapterImpl that provides a specific aspect
+	// This interface is extended by specific aspects
+	Adapter interface {
+		io.Closer
+		// Name returns the official name of this adapter. ex. "istio.io/statsd".
+		Name() string
+		// Description returns a user-friendly description of this adapter.
+		Description() string
+		// DefaultConfig returns a default configuration struct for this
+		// adapter Implmentation. This will be used by the configuration system to establish
+		// the shape of the block of configuration state passed to the NewAspect method.
+		DefaultConfig() (implConfig proto.Message)
+		// ValidateConfig determines whether the given configuration meets all correctness requirements.
+		ValidateConfig(implConfig proto.Message) error
+	}
+
+	// Aspect -- User visible cross cutting concern
+	Aspect interface {
+		io.Closer
+		// Name returns the official name of the aspect
+		Name() string
+	}
+
+	// Config is the internal struct used for the Aspect proto
+	// at present it only has an additional field that stores the converted
+	// and typed proto
+	Config struct {
+		istiopb.Aspect
+		// TypedParams points to the proto after google_protobuf.Struct is converted
+		TypedParams proto.Message
+	}
+
+	// AdapterConfig is the internal struct used for the Adapter proto
+	// at present it only has an additional field that stores the converted
+	// and typed Args
+	AdapterConfig struct {
```

Comment:
`aspect.AdapterConfig` is the internal version of
`istiopb.Adapter`

and
`aspect.Config` is the internal version of
`istiopb.Aspect`

will renaming `istiopb.Aspect --> istiopb.AspectConfig` and
`istiopb.Adapter --> istiopb.AdapterConfig` help ?

They are already in the `istio.config.v1` package.

or alternatively 
change
`istiopb "istio.io/api/istio/config/v1"`
to
`istioconfig "istio.io/api/istio/config/v1"`

```
	// Config (aspect.Config) is the internal struct used for the Aspect proto
	// at present it only has an additional field that stores the converted
	// and typed proto
	Config struct {
		istioconfig.Aspect
		// TypedParams points to the proto after google_protobuf.Struct is converted
		TypedParams proto.Message
	}
```